src/models/GitError.js

```diff
@@ -0,0 +1,147 @@
+  RefExistsError: `RefExistsError`,
+  NoHeadCommitError: `NoHeadCommitError`,
+  CommitNotFetchedError: `CommitNotFetchedError`,
+  ObjectTypeUnknownFail: `ObjectTypeUnknownFail`,
+  ObjectTypeAssertionFail: `ObjectTypeAssertionFail`,
+  ObjectTypeAssertionInTreeFail: `ObjectTypeAssertionInTreeFail`,
+  ObjectTypeAssertionInRefFail: `ObjectTypeAssertionInRefFail`,
+  ObjectTypeAssertionInPathFail: `ObjectTypeAssertionInPathFail`,
+  MissingAuthorError: `MissingAuthorError`,
+  GitRootNotFoundError: `GitRootNotFoundError`,
+  UnparseableServerResponseFail: `UnparseableServerResponseFail`,
+  InvalidDepthParameterError: `InvalidDepthParameterError`,
+  RemoteDoesNotSupportShallowFail: `RemoteDoesNotSupportShallowFail`,
+  RemoteDoesNotSupportDeepenSinceFail: `RemoteDoesNotSupportDeepenSinceFail`,
+  RemoteDoesNotSupportDeepenNotFail: `RemoteDoesNotSupportDeepenNotFail`,
+  RemoteDoesNotSupportDeepenRelativeFail: `RemoteDoesNotSupportDeepenRelativeFail`,
+  CorruptShallowOidFail: `CorruptShallowOidFail`,
+  FastForwardFail: `FastForwardFail`,
+  MergeNotSupportedFail: `MergeNotSupportedFail`,
+  DirectorySeparatorsError: `DirectorySeparatorsError`,
+  ResolveTreeError: `ResolveTreeError`,
+  DirectoryIsAFileError: `DirectoryIsAFileError`,
+  TreeOrBlobNotFoundError: `TreeOrBlobNotFoundError`,
+  NotImplementedFail: `NotImplementedFail`,
+  ReadObjectFail: `ReadObjectFail`,
+  ReadShallowObjectFail: `ReadShallowObjectFail`,
+  NotAnOidFail: `NotAnOidFail`,
+  NoRefspecConfiguredError: `NoRefspecConfiguredError`,
+  ResolveRefError: `ResolveRefError`,
+  ExpandRefError: `ExpandRefError`,
+  EmptyServerResponseFail: `EmptyServerResponseFail`,
+  AssertServerResponseFail: `AssertServerResponseFail`,
+  HTTPError: `HTTPError`,
+  RemoteUrlParseError: `RemoteUrlParseError`,
+  UnknownTransportError: `UnknownTransportError`,
+  AcquireLockFileFail: `AcquireLockFileFail`,
+  DoubleReleaseLockFileFail: `DoubleReleaseLockFileFail`,
+  InternalFail: `InternalFail`,
+  UnknownOauth2Format: `UnknownOauth2Format`,
+  MissingPasswordTokenError: `MissingPasswordTokenError`,
+  MissingUsernameError: `MissingUsernameError`,
+  MixPasswordTokenError: `MixPasswordTokenError`,
+  MixUsernamePasswordTokenError: `MixUsernamePasswordTokenError`,
+  MissingTokenError: `MissingTokenError`,
+  MixUsernameOauth2formatMissingTokenError: `MixUsernameOauth2formatMissingTokenError`,
+  MixPasswordOauth2formatMissingTokenError: `MixPasswordOauth2formatMissingTokenError`,
+  MixUsernamePasswordOauth2formatMissingTokenError: `MixUsernamePasswordOauth2formatMissingTokenError`,
+  MixUsernameOauth2formatTokenError: `MixUsernameOauth2formatTokenError`,
+  MixPasswordOauth2formatTokenError: `MixPasswordOauth2formatTokenError`,
+  MixUsernamePasswordOauth2formatTokenError: `MixUsernamePasswordOauth2formatTokenError`
+}
+
+export class GitError extends Error {
+  constructor (code, data) {
+    super()
+    this.name = code
+    this.code = code
```

Comment:
why two different fields with the same value?

Reply:
Ah! This one is interesting. The default `Error` class apparently uses the `.name` value in it's `.toString()` implementation. So by setting `.name` it magically starts showing up as "MyError: blaw blaw" instead of "Error: blaw blaw" in the console.

But... I felt like relying on `.name` was a little risky (ES6 will "automagically" assign `.name` to functions and classes in many cases. Try it `var bob = () => ''; bob.name`) so I stuck with `.code` as the "official" property to look at.